content/en/blog/_posts/2024-08-20-websocket-transition.md

```diff
@@ -0,0 +1,107 @@
+---
+layout: blog
+title: 'Streaming Transitions from SPDY to WebSockets'
+date: 2024-08-20
+slug: websockets-transition
+author: >
+  [Sean Sullivan](https://github.com/seans3) (Google)
+  [Shannon Kularathna](https://github.com/shannonxtreme) (Google)
+---
+
+In Kubernetes 1.31, by default kubectl now uses the WebSocket protocol
+instead of SPDY for streaming.
+
+This post describes what these changes mean for you and why these streaming APIs
+matter.
+
+## Streaming APIs in Kubernetes
+
+In Kubernetes, specific endpoints that are exposed as an HTTP or RESTful
+interface are upgraded to streaming connections, which require a streaming
+protocol. Unlike HTTP, which is a request-response protocol, a streaming
+protocol provides a persistent connection that's bi-directional, low-latency,
+and lets you interact in real-time. Streaming protocols support reading and
+writing data between your client and the server, in both directions, over the
+same connection. This type of connection is useful, for example, when you create
+a shell in a running container from your local workstation and run commands in
+the container.
+
+## Why the streaming protocol is changing
+
+Currently, Kubernetes uses the SPDY/3.1 protocol for upgraded streaming
+connections. SPDY/3.1 has been deprecated for eight years, and it was never
+standardized. Many modern proxies, gateways, and load balancers no longer
+support the protocol. As a result, you might notice that commands like
+`kubectl cp`, `kubectl attach`, `kubectl exec`, and `kubectl port-forward`
+stop working when you try to access your cluster through a proxy or gateway.
+
+As of Kubernetes 1.31, we're modifying the streaming protocol that Kubernetes
+uses for these commands to the more modern [WebSocket streaming protocol](https://datatracker.ietf.org/doc/html/rfc6455).
+The WebSocket protocol is a currently supported standardized streaming protocol
+that guarantees compatibility and interoperability with different components and
+programming languages. The WebSocket protocol is more widely supported by modern
+proxies and gateways than SPDY.
+
+## How streaming APIs work
+
+Kubernetes upgrades HTTP connections to streaming connections by adding
+specific upgrade headers to the originating HTTP request. For example, an HTTP
+upgrade request for running the `date` command on an `nginx` container within
+a cluster is similar to the following:
+
+```shell
```

Comment:
Don't mark this as shell as it's not a shell script. Try `console` instead.

Reply:
Fixed.